array-flatten.d.ts

```diff
@@ -1,16 +1,28 @@
-declare function flatten <T> (array: flatten.NestedArray<T>): T[];
+declare type InferElement<T extends ArrayLike<any>> = T extends IArguments
+  ? unknown
+  : {
+      array: T[number] extends infer Element
+        ? Element extends ArrayLike<any>
+          ? InferElement<Element>
+          : Element
+        : never
+      tuple: InferElement<Array<T[number]>>
+    }[Array<T[number]> extends T ? 'array' : 'tuple']
```

Comment:
What was the reasoning behind this switch between `tuple` and `array` behavior? I wasn't able to replicate the reason this was here when testing different solutions.

Reply:
I think support for type recursion was improved recently in TypeScript. Don't know the exact version off the top of my head

Reply:
3.7 included recursive types, but it didn't improve for within conditions - I was still getting the same error when testing this out. It's the main reason I still need to do:

```ts
{
  [K in Extract<keyof T, number>]: PickValue<T[K]>;
}[number]
```

As long as it wasn't any sort of explicit tuple support, that's fine - I made sure I have tests for tuples on `master` but wasn't sure if this did something I failed to test.

Reply:
Did you intentionally exclude the `arguments` support?

```ts
T extends IArguments
  ? unknown
  : // ...
```

Reply:
@aleclarson Not intentionally, it wasn't clear that line was there to just make it be the `unknown` type instead of `any`. I'm not sure how useful it is to make it `unknown` now that I think about it, but there's definitely a test for arguments on `master` that passes.

Reply:
`unknown` forces you to use type guards or an explicit typecast, so less runtime errors

Reply:
I’m aware of how it works. To clarify, I don’t see the point of changing a type on someone in this library if they’re using something unsafe - otherwise we could change all any types. I’d hope arguments usage is an edge case in TS, and by putting it in this library it’s condoning the usage arguments as the input type - which I believe it shouldn’t be used as for performance reasons.